File: src/Other/AddBinary.java
package Other;

// https://leetcode.com/problems/add-binary/
public class AddBinary {
    public String addBinary(String a, String b) {
        StringBuilder builder = new StringBuilder();
        int i = a.length() - 1, j = b.length() - 1, carry = 0;
        while (i >= 0 || j >= 0) {
            int sum = carry;
            if (i >= 0) {
                sum += a.charAt(i--) - '0';
            }
            if (j >= 0) {
                sum += b.charAt(j--) - '0';
            }
            builder.append(sum % 2);
            carry = sum / 2;
        }
        if (carry > 0) {
            builder.append(carry);
        }

        return builder.reverse().toString();
    }

    public static void main(String[] args) {
        AddBinary add = new AddBinary();
        String a = "1010";
        String b = "1011";
        System.out.println("Sum of binary: " + add.addBinary(a, b));
    }
}

File: py/essential/AddBinary.py
# https://leetcode.com/problems/add-binary/
# Math
# Grind


class Solution:
    def addBinary(self, a: str, b: str) -> str:
        res, carry, m, n = [], 0, len(a) - 1, len(b) - 1

        while m >= 0 or n >= 0:
            if m >= 0:
                carry += int(a[m])
                m -= 1
            if n >= 0:
                carry += int(b[n])
                n -= 1
            res.append(carry % 2)
            carry = carry // 2
        if carry > 0:
            res.append(carry)
        return "".join(str(num) for num in res[::-1])


s = Solution()
print(s.addBinary("11", "1"))
